### DISCUSSION TITLE
Extracting agreementRef from IPMode?

### DISCUSSION THREAD

**Discussion by MakakWasTaken** [phax/phase4#169] [Q&A]
I am currently using a custom implementation of `IAS4PModeAwareCryptoFactory` to get the ids of the `initiator` and `responder`. I the use these to determine what certificates to use. I would however like to use the agreementRef instead as it also contains the version of the certificate that should be used. 
I have looked at the entire IPMode interface and have not been able to find anything. The getter `getAgreement()` returns `urn:as4:agreement`

An example of an agreementRef with versioning looks like this:
`http://entsog.eu/communication/agreements/EIC_A/EIC_B/V` where `EIC_A`, `EIC_B` are the party’s EIC codes in alphabetical order and `V` is a number which increments which each certificate update.

**Comment by phax** [comment]
Lets try to reverse engineer the meaning of `agreement` and `agreementRef`

ebMS Core 3.0 spec:
> 1172 Agreeing on a P-Mode operation set is essential for two parties in order for their MSHs to interoperate.
> 1173 P-Modes are the MSH-level expression of a prior agreement between partners. A reference to such an
> 1174 agreement may be present in the message header (see eb:AgreementRef element in Section 5.2.2.7).
and
> 1647 AgreementRef is a string value that identifies the agreement that governs the exchange. The P-Mode
> 1648 under which the MSH operates for this message should be aligned with this agreement.
> 1649 The value of an AgreementRef element MUST be unique within a namespace mutually agreed by the two
> 1650 parties. This could be a concatenation of the From and To PartyId values, a URI containing the Internet
> 1651 domain name of one of the parties, or a namespace offered and managed by some other naming or
> 1652 registry service. It is RECOMMENDED that the AgreementRef be a URI.

The `agreementRef` is something that is soley contained in the UserMessage and not in the PMode. See below example (for the Peppol AS4 profile):
```xml
        <eb:CollaborationInfo>
          <eb:AgreementRef>urn:fdc:peppol.eu:2017:agreements:tia:ap_provider</eb:AgreementRef>
          <eb:Service type="cenbii-procid-ubl">urn:fdc:peppol.eu:2017:poacc:billing:01:1.0</eb:Service>
          <eb:Action>busdox-docid-qns::urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice##urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0::2.1</eb:Action>
          <eb:ConversationId>phase4@Conv7764093634010523874</eb:ConversationId>
        </eb:CollaborationInfo>
```

I think what you need is first a custom `IPModeResolver` implementation that lets you select a `PMode` based on the `agreementRef` value.

Now my detailed knowledge about ENTSOG is de facto zero. Do you need point-to-point PModes or do you use "general PModes" that only differ by certificate(s) and endpoint URL? In the latter case, you could define *n* PModes - one for each agreementRef....

hth
